### PR TITLE
added jwk

### PR DESCRIPTION
**Pull request by sabify**
Hello,

This PR should be considered a draft until we have stable `k256` and `p256` with `jwk` feature. #14

### CODE REVIEW

**Comment by sabify** [comment]
This PR is ready to be merged.

File: src/algorithms/es256.rs [hunk 1]
```diff
@@ -26,6 +27,28 @@ impl P256PublicKey {
...
+
+    pub fn to_jwk(&self) -> JwkEcKey {
+        let pk = PublicKey::<p256::NistP256>::from(&self.0);
+        pk.to_jwk()
+    }
+
+    pub fn to_jwk_str(&self) -> String {
+        let pk = PublicKey::<p256::NistP256>::from(&self.0);
+        pk.to_jwk_string()
+    }
```
Comment (tarcieri) [inline src/algorithms/es256.rs:50]: Hmm, this makes me think I should add a `jwk` feature to the `ecdsa` crate...

**Comment by jedisct1** [comment]
The vast majority of keys seem to be either SHA2 keys or RSA keys. In fact, I've never seen any `jwks.json` file with something else so far :(

Objects containing multiple keys also seem to be very common (random example: Auth0). So, in the context of that crate, we may want to store all the received keys that are valid and match the expected algorithm. And, later, transparently use the correct one based on the key id, while still handling the case where there is no key id.

File: src/algorithms/es256.rs (hunk 1 above)
Comment (sabify) [reply src/algorithms/es256.rs:50]: yeah, that could be great!

File: src/algorithms/es256.rs [hunk 2]
```diff
@@ -26,6 +27,28 @@ impl P256PublicKey {
...
+        Ok(P256PublicKey(pk))
+    }
+
+    pub fn from_jwk(jwk: &JwkEcKey) -> Result<Self, Error> {
+        let pk = PublicKey::from_jwk(jwk).map_err(|_| JWTError::InvalidPublicKey)?;
+        let pk = ecdsa::VerifyingKey::from(pk);
+        Ok(P256PublicKey(pk))
+    }
+
+    pub fn to_jwk(&self) -> JwkEcKey {
```
[the remark this answers is not in the document]
Comment (sabify) [reply src/algorithms/es256.rs:42]: Actually, removing a feature won't necessarily make the API simple, but renaming the methods to be more convenient to the most of users is great idea and also can keep the consistency of the API, specially in future development of other algorithms to have `JWK` feature.

`JwkEcKey` is already exposed and having this feature may be useful to other niche users...

**Comment by sabify** [comment]
> The vast majority of keys seem to be either SHA2 keys or RSA keys. In fact, I've never seen any `jwks.json` file with something else so far :(
> 
> Objects containing multiple keys also seem to be very common (random example: Auth0). So, in the context of that crate, we may want to store all the received keys that are valid and match the expected algorithm. And, later, transparently use the correct one based on the key id, while still handling the case where there is no key id.

You are right, this crate can have a separate helper module to handle all `jwk` and `jwks` cases conveniently and that could be great!

**Comment by aweis89** [comment]
Any chance we can get this merged?

**Comment by jedisct1** [comment]
> Any chance we can get this merged?

There's still a lot left to be done :)

**Comment by sabify** [comment]
So, what happened to this PR?
You don't like to accept small changes and improvements to your repo? just want all sort of jwks to be handled and added at once and wait for it?

**Comment by jedisct1** [comment]
Hi!

Adding JWKS support would be great!

But maybe this should happen in a separate branch until it is ready. Merging work in progress blocks new releases or can lead to breaking changes later when other parts are implemented.

**Comment by sabify** [comment]
Hi there!

I don't think this PR and the feature for those few algorithms are in WIP/draft and not even related to what you've mentioned. The api is clear and the jwk feature for those algorithms is implemented and we're just exposing them to users.

What you've mentioned is having a separate jwk module (another layer) on top of these exposed internal methods to handle all sort of jwks conveniently which is not applicable right now because we need to wait for crypto group to add those apis.

So, we need these small changes before implementing the desired module.

**Comment by jedisct1** [comment]
What people are given to use as a verification key is a URL like this: https://samples.auth0.com/.well-known/jwks.json

So, there should be a `RSAPublicKey::from_jwks()` function that accepts what they get from this URL. The fact that this encodes a set of JWKs is an internal detail that most users shouldn't have to have to deal with.

**Pull request closed without merging**